27сен23

# Продолжение Pandas. CSV, DataFrame.
* Как создать DataFrame?
* Почему у DataFrame самая удобная индексация?
* Что такое csv файл и почему он так называется?
* Откуда мне взять csv файл и как открыть его в питоне?
* Какие методы есть у обьекта DataFrame?

## Создание и индексация DataFrame

Полная документация по [**ссылке**](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html).

В общем-то DataFrame – самая обычная таблица, многомерный массив (не обязательно двумерный). В этом курсе будут в основном двумерные таблицы, так как я убежден, что в любом типе данных самое главное это не количество измерений, а умение пользоваться его методами.

In [1]:
from pandas import DataFrame
df = DataFrame({'A' : [1, 2, 3],
                'B' : [4, 5, 6],
                'C' : [7, 8, 9]})
df.index = [100, 200, 300]
df

A  B  C
100  1  4  7
200  2  5  8
300  3  6  9

Индексация DataFrame предоставляет пользователю практически неограниченные возможности.

In [2]:
# можно дать новые значения всему столбцу
df['A'] = [10, 11, 12]
df

A  B  C
100  10  4  7
200  11  5  8
300  12  6  9

In [3]:
# можно дать новые значения всей строке по индексу строки
df.loc[200] = [13, 14, 15]
df

A   B   C
100  10   4   7
200  13  14  15
300  12   6   9

In [4]:
# можно дать новые значения всей строке по номеру строки
df.iloc[0] = [16, 17, 18]
df

A   B   C
100  16  17  18
200  13  14  15
300  12   6   9

In [5]:
# можно дать новое значение элементу по обоим индексам
df.at[300, 'C'] = 19
df

A   B   C
100  16  17  18
200  13  14  15
300  12   6  19

In [6]:
# можно дать новые значения элементам по нескольким координатам
df.loc[[200, 300], 'A'] = 20
df

A   B   C
100  16  17  18
200  20  14  15
300  20   6  19

In [7]:
# можно дать новые значения элементам по какому-либо условию
df.loc[df['B'] > 10, 'C'] = 21
df

A   B   C
100  16  17  21
200  20  14  21
300  20   6  19

## Создаем DataFrame из *.csv*
Как пишет [**Википедия**](https://ru.wikipedia.org/wiki/CSV), "CSV" расшифровывается как "Comma-Separated Values", что (на случай, если ты не знаешь английского) означает "значения, разделенные запятой". В названии все сказано, csv файл это просто набор значений, записанных через запятую.

В этом курсе для чтения csv файлов мы будем использовать функцию *read_csv()* из библиотеки pandas. Она создает обьект типа DataFrame и кладет туда все значения из файлов.

In [8]:
# так можно загрузить csv в питон
from pandas import read_csv

url = "https://raw.githubusercontent.com/MainakRepositor/Datasets/master/anime.csv"

# здесь я указал, по какому из столбцов индексировать DataFrame
anime = read_csv(url, index_col="anime_id")

# так можно вывести первые несколько строк
anime.head(10)

name  \
anime_id                                                      
32281                                        Kimi no Na wa.   
5114                       Fullmetal Alchemist: Brotherhood   
28977                                              Gintama°   
9253                                            Steins;Gate   
9969                                          Gintama&#039;   
32935     Haikyuu!!: Karasuno Koukou VS Shiratorizawa Ga...   
11061                                Hunter x Hunter (2011)   
820                                    Ginga Eiyuu Densetsu   
15335     Gintama Movie: Kanketsu-hen - Yorozuya yo Eien...   
15417                              Gintama&#039;: Enchousen   

                                                      genre   type episodes  \
anime_id                                                                      
32281                  Drama, Romance, School, Supernatural  Movie        1   
5114      Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64   
28977     Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
9253                                       Sci-Fi, Thriller     TV       24   
9969      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
32935                Comedy, Drama, School, Shounen, Sports     TV       10   
11061               Action, Adventure, Shounen, Super Power     TV      148   
820                          Drama, Military, Sci-Fi, Space    OVA      110   
15335     Action, Comedy, Historical, Parody, Samurai, S...  Movie        1   
15417     Action, Comedy, Historical, Parody, Samurai, S...     TV       13   

          rating  members  
anime_id                   
32281       9.37   200630  
5114        9.26   793665  
28977       9.25   114262  
9253        9.17   673572  
9969        9.16   151266  
32935       9.15    93351  
11061       9.13   425855  
820         9.11    80679  
15335       9.10    72534  
15417       9.11    81109

Сам csv файл я никак не редактировал и даже не сохранял себе на компьютер. Он лежит по ссылке, которая записана в переменной *url*.

Далее некоторые функции и методы, применимые к обьекту DataFrame.

In [9]:
# так можно узнать количество строк в DataFrame
len(anime)

12294

Так вот у нас 12294 строки, в каждой есть id этого аниме, его название, его жанры, тип, поличество эпизодов, рейтиг и "*members*", что бы это не значило. 

Для начала отсортирую по "*anime_id*", так как будто приятнее.

In [10]:
# сортировка по индексу
anime = anime.sort_index()
anime

name  \
anime_id                                                      
1                                              Cowboy Bebop   
5                           Cowboy Bebop: Tengoku no Tobira   
6                                                    Trigun   
7                                        Witch Hunter Robin   
8                                    Beet the Vandel Buster   
...                                                     ...   
34514                                   Pokemon Generations   
34519                   Mobile Suit Gakuen: G-Reco Koushien   
34522                             Wake Up, Girls! Shin Shou   
34525                                     Centaur no Nayami   
34527     Gou-chan. Moko to Chinjuu no Mori no Nakama-tachi   

                                                      genre     type episodes  \
anime_id                                                                        
1           Action, Adventure, Comedy, Drama, Sci-Fi, Space       TV       26   
5                     Action, Drama, Mystery, Sci-Fi, Space    Movie        1   
6                                    Action, Comedy, Sci-Fi       TV       26   
7         Action, Drama, Magic, Mystery, Police, Superna...       TV       26   
8                 Adventure, Fantasy, Shounen, Supernatural       TV       52   
...                                                     ...      ...      ...   
34514                Action, Adventure, Fantasy, Game, Kids      ONA       18   
34519                                                Comedy  Special        9   
34522                                          Drama, Music       TV  Unknown   
34525          Comedy, Fantasy, Slice of Life, Supernatural       TV  Unknown   
34527                                       Adventure, Kids    Movie        1   

          rating  members  
anime_id                   
1           8.82   486824  
5           8.40   137636  
6           8.32   283069  
7           7.36    64905  
8           7.06     9848  
...          ...      ...  
34514       7.21      295  
34519       5.67       94  
34522        NaN      381  
34525        NaN      108  
34527        NaN        5  

[12294 rows x 6 columns]

Так он выводит первые и последние 5 строчек, а так же размеры. Часто в некоторых местах встречается значение NaN, если ты забыл, что это – чекни [**предыдущий урок**](https://github.com/teacher57/data_analysis_course/blob/main/2nd%20Lesson.ipynb).

In [11]:
from pandas import Series
from numpy import NaN

# нужно создать массив с такими же индексами, как названия столбцов DataFrame
new_row = Series({"anime_id" : 15000,
                  "name"     : "BEST_ANIME_EVER",
                  "genre"    : NaN,
                  "type"     : "TV",
                  "episodes" : 57,
                  "rating"   : 10,
                  "members"  : 57000000})

# добавление строки
anime.loc[34528] = new_row
anime

name  \
anime_id                                                      
1                                              Cowboy Bebop   
5                           Cowboy Bebop: Tengoku no Tobira   
6                                                    Trigun   
7                                        Witch Hunter Robin   
8                                    Beet the Vandel Buster   
...                                                     ...   
34519                   Mobile Suit Gakuen: G-Reco Koushien   
34522                             Wake Up, Girls! Shin Shou   
34525                                     Centaur no Nayami   
34527     Gou-chan. Moko to Chinjuu no Mori no Nakama-tachi   
34528                                       BEST_ANIME_EVER   

                                                      genre     type episodes  \
anime_id                                                                        
1           Action, Adventure, Comedy, Drama, Sci-Fi, Space       TV       26   
5                     Action, Drama, Mystery, Sci-Fi, Space    Movie        1   
6                                    Action, Comedy, Sci-Fi       TV       26   
7         Action, Drama, Magic, Mystery, Police, Superna...       TV       26   
8                 Adventure, Fantasy, Shounen, Supernatural       TV       52   
...                                                     ...      ...      ...   
34519                                                Comedy  Special        9   
34522                                          Drama, Music       TV  Unknown   
34525          Comedy, Fantasy, Slice of Life, Supernatural       TV  Unknown   
34527                                       Adventure, Kids    Movie        1   
34528                                                   NaN       TV       57   

          rating   members  
anime_id                    
1           8.82    486824  
5           8.40    137636  
6           8.32    283069  
7           7.36     64905  
8           7.06      9848  
...          ...       ...  
34519       5.67        94  
34522        NaN       381  
34525        NaN       108  
34527        NaN         5  
34528      10.00  57000000  

[12295 rows x 6 columns]

In [12]:
# добавление столбца
anime["col"] = [NaN] * 12295
anime.head(5)

name  \
anime_id                                    
1                            Cowboy Bebop   
5         Cowboy Bebop: Tengoku no Tobira   
6                                  Trigun   
7                      Witch Hunter Robin   
8                  Beet the Vandel Buster   

                                                      genre   type episodes  \
anime_id                                                                      
1           Action, Adventure, Comedy, Drama, Sci-Fi, Space     TV       26   
5                     Action, Drama, Mystery, Sci-Fi, Space  Movie        1   
6                                    Action, Comedy, Sci-Fi     TV       26   
7         Action, Drama, Magic, Mystery, Police, Superna...     TV       26   
8                 Adventure, Fantasy, Shounen, Supernatural     TV       52   

          rating  members  col  
anime_id                        
1           8.82   486824  NaN  
5           8.40   137636  NaN  
6           8.32   283069  NaN  
7           7.36    64905  NaN  
8           7.06     9848  NaN

Удаление чего-либо из таблицы – вообще одно удовольствие, потому что в pandas одна функция позволяет удалять строки и столбцы, в зависимости от указанных параметров.

Параметр *axis* отвечает за то, по какой оси будет происходить поиск указанного индекса. Параметр *inplace* позволяет произвести изменение в таблице, без присвоения.

In [13]:
# удаление строки с помощью метода drop
anime.drop(34528, axis=0, inplace=True)

# так можно вывести несколько последних элементов
anime.tail(5)

name  \
anime_id                                                      
34514                                   Pokemon Generations   
34519                   Mobile Suit Gakuen: G-Reco Koushien   
34522                             Wake Up, Girls! Shin Shou   
34525                                     Centaur no Nayami   
34527     Gou-chan. Moko to Chinjuu no Mori no Nakama-tachi   

                                                 genre     type episodes  \
anime_id                                                                   
34514           Action, Adventure, Fantasy, Game, Kids      ONA       18   
34519                                           Comedy  Special        9   
34522                                     Drama, Music       TV  Unknown   
34525     Comedy, Fantasy, Slice of Life, Supernatural       TV  Unknown   
34527                                  Adventure, Kids    Movie        1   

          rating  members  col  
anime_id                        
34514       7.21      295  NaN  
34519       5.67       94  NaN  
34522        NaN      381  NaN  
34525        NaN      108  NaN  
34527        NaN        5  NaN

Если с помощью метода *.drop()* попытаться удалить элемент, которого нет в таблице – питон выдаст ошибку.

In [14]:
# удаление столбца (по названию)
anime.drop("col", axis=1, inplace=True)
anime.head(5)

name  \
anime_id                                    
1                            Cowboy Bebop   
5         Cowboy Bebop: Tengoku no Tobira   
6                                  Trigun   
7                      Witch Hunter Robin   
8                  Beet the Vandel Buster   

                                                      genre   type episodes  \
anime_id                                                                      
1           Action, Adventure, Comedy, Drama, Sci-Fi, Space     TV       26   
5                     Action, Drama, Mystery, Sci-Fi, Space  Movie        1   
6                                    Action, Comedy, Sci-Fi     TV       26   
7         Action, Drama, Magic, Mystery, Police, Superna...     TV       26   
8                 Adventure, Fantasy, Shounen, Supernatural     TV       52   

          rating  members  
anime_id                   
1           8.82   486824  
5           8.40   137636  
6           8.32   283069  
7           7.36    64905  
8           7.06     9848

Без параметра *inplace* метод *.drop()* не поменяет саму таблицу. Без него можно использовать конструкцию:
```python
anime = anime.drop("col", axis=1)
```


Причина, по которой мы на прошлом уроке проходили Series, заключается в том, что выделение одного столбца или строки из DataFrame выдает именно обьект типа Series.

In [15]:
print(type(anime["rating"]))
anime["rating"]

<class 'pandas.core.series.Series'>


anime_id
1        8.82
5        8.40
6        8.32
7        7.36
8        7.06
         ... 
34514    7.21
34519    5.67
34522     NaN
34525     NaN
34527     NaN
Name: rating, Length: 12294, dtype: float64

In [16]:
print(type(anime.loc[1]))
anime.loc[1]

<class 'pandas.core.series.Series'>


name                                           Cowboy Bebop
genre       Action, Adventure, Comedy, Drama, Sci-Fi, Space
type                                                     TV
episodes                                                 26
rating                                                 8.82
members                                              486824
Name: 1, dtype: object

In [17]:
# так можно узнать размеры таблицы
anime.shape

(12294, 6)

In [18]:
# а вот так можно получить краткую статистическую информацию о численных столбцах таблицы
anime.describe()

rating       members
count  12064.000000  1.229400e+04
mean       6.473902  1.807134e+04
std        1.026746  5.482068e+04
min        1.670000  5.000000e+00
25%        5.880000  2.250000e+02
50%        6.570000  1.550000e+03
75%        7.180000  9.437000e+03
max       10.000000  1.013917e+06

* *mean* – среднее арифметическое
* *std*  – стандартное отклонение


В задачах для самостоятельного выполнения пригодится знание еще пары фишек.

### 1. Remove

In [19]:
anime.tail(5)

name  \
anime_id                                                      
34514                                   Pokemon Generations   
34519                   Mobile Suit Gakuen: G-Reco Koushien   
34522                             Wake Up, Girls! Shin Shou   
34525                                     Centaur no Nayami   
34527     Gou-chan. Moko to Chinjuu no Mori no Nakama-tachi   

                                                 genre     type episodes  \
anime_id                                                                   
34514           Action, Adventure, Fantasy, Game, Kids      ONA       18   
34519                                           Comedy  Special        9   
34522                                     Drama, Music       TV  Unknown   
34525     Comedy, Fantasy, Slice of Life, Supernatural       TV  Unknown   
34527                                  Adventure, Kids    Movie        1   

          rating  members  
anime_id                   
34514       7.21      295  
34519       5.67       94  
34522        NaN      381  
34525        NaN      108  
34527        NaN        5

Вспомним, что в таблице есть значения "Unknown" в столбце "*episodes*".

Допустим, нам нужно заменить эти значения на NaN.

In [20]:
anime["episodes"].replace("Unknown", NaN, inplace=True)
anime.tail(5)

name  \
anime_id                                                      
34514                                   Pokemon Generations   
34519                   Mobile Suit Gakuen: G-Reco Koushien   
34522                             Wake Up, Girls! Shin Shou   
34525                                     Centaur no Nayami   
34527     Gou-chan. Moko to Chinjuu no Mori no Nakama-tachi   

                                                 genre     type episodes  \
anime_id                                                                   
34514           Action, Adventure, Fantasy, Game, Kids      ONA       18   
34519                                           Comedy  Special        9   
34522                                     Drama, Music       TV      NaN   
34525     Comedy, Fantasy, Slice of Life, Supernatural       TV      NaN   
34527                                  Adventure, Kids    Movie        1   

          rating  members  
anime_id                   
34514       7.21      295  
34519       5.67       94  
34522        NaN      381  
34525        NaN      108  
34527        NaN        5

И это не единственное, что может сделать этот метод. Вот еще один пример его возможностей.

In [21]:
anime["type"].replace(["ONA", "TV"], ["Original", "Television"], inplace=True)
anime.tail(5)

name  \
anime_id                                                      
34514                                   Pokemon Generations   
34519                   Mobile Suit Gakuen: G-Reco Koushien   
34522                             Wake Up, Girls! Shin Shou   
34525                                     Centaur no Nayami   
34527     Gou-chan. Moko to Chinjuu no Mori no Nakama-tachi   

                                                 genre        type episodes  \
anime_id                                                                      
34514           Action, Adventure, Fantasy, Game, Kids    Original       18   
34519                                           Comedy     Special        9   
34522                                     Drama, Music  Television      NaN   
34525     Comedy, Fantasy, Slice of Life, Supernatural  Television      NaN   
34527                                  Adventure, Kids       Movie        1   

          rating  members  
anime_id                   
34514       7.21      295  
34519       5.67       94  
34522        NaN      381  
34525        NaN      108  
34527        NaN        5

*inplace* в этом случае значит то же, что и в методе *.drop()*

## 2. NaN != NaN

In [22]:
from numpy import nan
nan == nan

False

NaN не равен сам себе. Так что, чтобы проверить является ли значение NaN используется:

In [23]:
from numpy import isnan
if isnan(nan):
    print("is nan")
    

is nan


---
## Задачи для самостоятельного выполнения
Дедлайн 3окт23 23:59.

|Вариант 1|Вариант 2|
|:-----------:|:----------:|
|CSV файл по [ссылке](https://raw.githubusercontent.com/toUpperCase78/formula1-datasets/master/Formula1_2023season_raceResults.csv)|CSV файл по [ссылке](https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv)|
|1. На каждом треке найти лучшего гонщика и его время (вывести всех одним фреймом)|1. В каждом классе найти самого молодого и самого старого выжившего пассажира (вывести всех одним фреймом)|
|2. Найти среднее время на каждом треке (только среди тех, кто доехал до финиша)|2. Посчитать количество выживших в каждом классе|
|3. Посчитать (программно, разумеется) сколько раз Верстаппен обогнал Гамильтона, а сколько – наоборот |3. Заменить неуказанный возраст у тех, кто не выжил на "В полном расцвете сил" (вывести только их одним фреймом)|

Не забудь, что домашку нужно присылать в формате .ipynb